### PR TITLE
Reserve space based on size_hint prior to pushing.

### PR DESCRIPTION
**Pull request by hbina**
Not sure how much this will effect. Is there any benchmark that can be run? I assume this is _almost always_ a good thing

### CODE REVIEW

File: src/cargo/util/dependency_queue.rs [hunk 1]
```diff
@@ -35,13 +35,13 @@ pub struct DependencyQueue<N: Hash + Eq, E: Hash + Eq, V> {
...
 
-impl<N: Hash + Eq, E: Hash + Eq, V> Default for DependencyQueue<N, E, V> {
+impl<N: Hash + Eq + Clone, E: Hash + Eq + Clone, V> Default for DependencyQueue<N, E, V> {
     fn default() -> DependencyQueue<N, E, V> {
         DependencyQueue::new()
     }
 }
 
-impl<N: Hash + Eq, E: Hash + Eq, V> DependencyQueue<N, E, V> {
+impl<N: Hash + Eq + Clone, E: Eq + Hash + Clone, V> DependencyQueue<N, E, V> {
```
Comment (hbina) [inline src/cargo/util/dependency_queue.rs:44]: Oops, I think these are unrelated changes...

**Comment by ehuss** [comment]
There isn't an explicit set of benchmarks for this sort of thing.  Typically what I do is run `cargo check` a few times in a large project. I like to use [hyperfine](https://crates.io/crates/hyperfine) for this, something like `hyperfine --warmup 1 "cargo check"`.  

I suspect this won't have any measurable impact, and generally I'd prefer not to add additional code/complexity unless there is a good motivating reason.  But give it a shot and see if it helps!
